### PR TITLE
Changed search result ad clicked URL

### PR DESCRIPTION
**Pull request by aseren**
Changed search result ad clicked URL to:
https://search.brave.com/a/redirect?click_url=XXXXXX&placement_id=XXXXXXX&creative_instance_id=XXXXXXX

Resolves https://github.com/brave/brave-browser/issues/28252

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Test case 1
- Turn off Brave Private Ads.
- Do several search requests on search.brave.com until an ad is shown in a SERP.
- Make sure that a search result ad viewed event is not triggered for Brave Ads.
- Click on the ad link.
- Make sure that search result ad clicked event is not triggered for Brave Ads.

### Test case 2
- Turn on Brave Private Ads.
- Do several search requests on search.brave.com until an ad is shown in a SERP.
- Make sure that a search result ad viewed event is triggered. There should be `Viewed search result ad with placement id` message in rewards-internals log.
- Click on the ad link.
- Make sure that a search result ad clicked event is triggered. There should be `Clicked search result ad with placement id` message in rewards-internals log.

### CODE REVIEW

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc [hunk 1]
```diff
@@ -16,12 +16,16 @@
...
 #include "brave/vendor/bat-native-ads/include/bat/ads/public/interfaces/ads.mojom.h"
+#include "url/gurl.h"
 
 using ads::mojom::ConversionInfo;
 using ads::mojom::ConversionInfoPtr;
 using ads::mojom::SearchResultAdInfo;
 using ads::mojom::SearchResultAdInfoPtr;
 
+using SearchResultAdMap =
+    base::flat_map<std::string, ads::mojom::SearchResultAdInfoPtr>;
```
Comment (tmancey) [inline components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc:27]: nit: comment for first param, see last comment

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc [hunk 2]
```diff
@@ -270,8 +274,7 @@ void LogSearchResultAdMap(const SearchResultAdMap& search_result_ads) {
     return;
   }
 
-  for (const auto& search_result_ad_pair : search_result_ads) {
-    const auto& search_result_ad = search_result_ad_pair.second;
+  for (const auto& [placement_id, search_result_ad] : search_result_ads) {
```
Comment (tmancey) [inline components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc:277]: What does it mean when an ad is converted?

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "url/gurl.h"
+#include "url/third_party/mozilla/url_parse.h"
+
+namespace brave_ads {
+
+namespace {
+
+constexpr char kSearchResultAdClickedPath[] = "/a/redirect";
+constexpr char kPlacementIdParameterName[] = "placement_id";
```
Comment (tmancey) [inline components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc:18]: QueryKey?

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+absl::optional<std::string> GetPlacementIdFromSearchResultAdClickedUrl(
+    const GURL& url) {
+  if (!url.is_valid() || !url.SchemeIs(url::kHttpsScheme) ||
+      url.path_piece() != kSearchResultAdClickedPath || !url.has_query() ||
+      !brave_search::IsAllowedHost(url)) {
+    return {};
+  }
+
+  base::StringPiece query_str = url.query_piece();
```
Comment (tmancey) [inline components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc:30]: Do we need _str

**Review by tmancey — APPROVED** [review]
LGTM (see nits)

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc (hunk 2 above)
Comment (aseren) [reply components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc:277]: This means, that an ad was converted from JSON-ld format to SearchResultAdPtr struct. `Converted` word can be confusing here. I will remove it.

File: components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.h [hunk 5]
```diff
@@ -57,7 +58,7 @@ class SearchResultAdHandler final {
   raw_ptr<AdsService> ads_service_ = nullptr;  // NOT OWNED
   bool should_trigger_viewed_event_ = true;
 
-  absl::optional<base::flat_map<GURL, ads::mojom::SearchResultAdInfoPtr>>
+  absl::optional<base::flat_map<std::string, ads::mojom::SearchResultAdInfoPtr>>
```
[the remark this answers is not in the document]
Comment (tmancey) [reply components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.h:61]: May not be needed here but we should def add comment for declaration

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc:277]: Maybe JsonLdToSearchResultAd

Comment (aseren) [reply components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.cc:277]: changed to `A search result ad was seen`

File: components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (aseren) [reply components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.h:61]: I think this comment is good here.

File: components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc (hunk 4 above)
Comment (aseren) [reply components/brave_ads/core/browser/search_result_ad/search_result_ad_util.cc:30]: Not needed actually. Refactored naming.

**Comment by aseren** [comment]
Windows build failed due to CI issues, all tests passed.
https://bravesoftware.slack.com/archives/C7VLGSR55/p1676397869997049

**Comment by btlechowski** [comment]
Good to uplift. The functionality is behind a flag which is disabled by default so it is low risk.

Verification passed on

Brave | 1.50.39 Chromium: 110.0.5481.100 (Official Build) nightly (64-bit) 
-- | --
Revision | 4be7a36f7cb943af6118e449bbab494b43dcaddd-refs/branch-heads/5481_77@{#14}
OS | Ubuntu 18.04 LTS

Verified the above test plan

#### Test case 1 - ads disabled

<details>

Verified search ad is shown
image


Verified in the log search ad is not processed at all


</details>


#### Test case 2 - ads enabled

<details>
Note: run Brave with following command line to enable functionality
```
--enable-logging=stderr --vmodule="*/variations/*"=6,"*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6,"*/bat_ads/*"=6 --enable-features=SupportBraveSearchResultAdConfirmationEvents

```


Verified search ad is shown

```
[2567:2567:0216/145554.762247:VERBOSE6:search_result_ad_converting_util.cc(308)] A search result ad was delivered:
  "data-placement-id": 97a97811-2adc-455b-b7d4-16007da9e334
  "data-creative-instance-id": 53e5f825-25e3-4ad7-8331-536bae478d77
  "data-creative-set-id": aa563bd4-db73-4f6e-8d08-ce0c8b5155c7
  "data-campaign-id": 7c1178a6-20a8-4223-bffd-f1909f719393
  "data-advertiser-id": 1f981714-582b-40b5-95f4-3181b7042afe
  "data-landing-page": https://register.sandbox.game/bravesearch?utm_source=brave&utm_medium=display&utm_campaign=bravesearchNFT
  "data-headline-text": The Sandbox - Official Site
  "data-description": Own Land, Build and Play games, and Earn on your creations in the Metaverse. Join Today!
  "data-rewards-value": 0

```

Verified `served` event is processed
```
[2567:2567:0216/145554.842611:VERBOSE3:search_result_ad_event_served.cc(19)] Served search result ad with placement id 97a97811-2adc-455b-b7d4-16007da9e334 and creative instance id 53e5f825-25e3-4ad7-8331-536bae478d77
[2567:2567:0216/145554.863192:VERBOSE6:search_result_ad_event_served.cc(28)] Successfully logged search result ad served event

```

Verified `view` event is processed
```
[2567:2567:0216/145554.894809:VERBOSE3:search_result_ad_event_viewed.cc(19)] Viewed search result ad with placement id 97a97811-2adc-455b-b7d4-16007da9e334 and creative instance id 53e5f825-25e3-4ad7-8331-536bae478d77
[2567:2567:0216/145554.915150:VERBOSE6:search_result_ad_event_viewed.cc(28)] Successfully logged search result ad viewed event

```

Verified `click` event is processed
```
[2567:2567:0216/145559.196842:VERBOSE3:search_result_ad_event_clicked.cc(19)] Clicked search result ad with placement id 97a97811-2adc-455b-b7d4-16007da9e334 and creative instance id 53e5f825-25e3-4ad7-8331-536bae478d77
[2567:2567:0216/145559.264380:VERBOSE6:search_result_ad_event_clicked.cc(28)] Successfully logged search result ad clicked event

```

Verified going back in history does not trigger `view` event again, got only log that search result ad was delivered
```
[3411:3411:0216/151740.800065:VERBOSE6:search_result_ad_converting_util.cc(308)] A search result ad was delivered:
  "data-placement-id": 52e68d88-01b2-48ae-9c6f-865851a7d8b4
  "data-creative-instance-id": 3a6ed048-392e-4f8a-9d8a-83338f5768b1
  "data-creative-set-id": dcbb167c-c667-48fe-acfb-d14a109022f7
  "data-campaign-id": 1d130998-b5ef-4258-8c96-5cf58b80f647
  "data-advertiser-id": 735a260f-45d5-4b29-b2f0-9918e910f0ae
  "data-landing-page": https://med.etoro.com/aw.aspx?B=20017&A=84346&Task=Click&SubAffiliateID=352042739
  "data-headline-text": eToro™ - Official Site
  "data-description": Discuss Real-Time Market Trends with Top Investors. 79% Of Retail CFD Accounts Lose Money
  "data-rewards-value": 0

```


</details>

**Pull request merged**